### PR TITLE
Create main developer guide for Python

### PR DESCRIPTION
**Pull request by vyasr**
This PR adds a primary developer guide for Python. It provides a more complete and informative landing page for new developers. When #11217, #11199, and #11122 are merged, they will all be linked from this page to provide a complete set of developer documentation.

There is one main point of discussion that I would like reviewer comments on, and that is the section on directory and file organization. How do we want that aspect of cuDF to look?

### CODE REVIEW

File: docs/cudf/source/developer_guide/developer_guide.md
Comment (bdice) [inline docs/cudf/source/developer_guide/developer_guide.md:59]: The mention of 22.06 is confusing. Maybe “if we decide to remove an API during the 22.08 development cycle, it will be deprecated in the 22.08 release and removed in the 22.10 release.”

File: docs/cudf/source/developer_guide/developer_guide.md
Comment (bdice) [inline docs/cudf/source/developer_guide/developer_guide.md:61]:
```suggestion
All internal usage of deprecated APIs in cuDF should be removed when the API is deprecated.
```

File: docs/cudf/source/developer_guide/developer_guide.md
Comment (bdice) [inline docs/cudf/source/developer_guide/developer_guide.md:62]:
```suggestion
This prevents unexpected warnings from surfacing when other (non-deprecated) APIs are called.
```

File: docs/cudf/source/developer_guide/developer_guide.md
Comment (bdice) [inline docs/cudf/source/developer_guide/developer_guide.md:65]:
```suggestion
When deprecating an API, developers should open a corresponding GitHub issue to track the API removal.
```

File: docs/cudf/source/developer_guide/developer_guide.md
Comment (bdice) [inline docs/cudf/source/developer_guide/developer_guide.md:66]:
```suggestion
The GitHub issue should be labeled "deprecation" and added to the next release's project board.
```

File: docs/cudf/source/developer_guide/developer_guide.md
Comment (bdice) [inline docs/cudf/source/developer_guide/developer_guide.md:72]: It’s not really a throw/raise because it doesn’t halt execution.
```suggestion
- Emit a `FutureWarning`.
```

File: docs/cudf/source/developer_guide/developer_guide.md
Comment (bdice) [inline docs/cudf/source/developer_guide/developer_guide.md:76]:
```suggestion
- Don't specify a version when removal will occur.
```

File: docs/cudf/source/developer_guide/developer_guide.md
Comment (bdice) [inline docs/cudf/source/developer_guide/developer_guide.md:90]: Can you cite where you’re pulling this from? I have seen the difference described differently. It’s not about whether the warning comes from Python’s internals, but whether the target audience is developers or application users. https://docs.python.org/3/library/warnings.html

**Review by bdice — COMMENTED** [review]
Nice! Here are some drive-by comments (I got excited to see your work on this).

File: docs/cudf/source/developer_guide/developer_guide.md
Comment (vyasr) [reply docs/cudf/source/developer_guide/developer_guide.md:90]: The best description is probably in the [original PEP that changed the recommended warnings](https://peps.python.org/pep-0565/#additional-use-case-for-futurewarning). The wording that you're referring to is more canonical (it's what shows up in the Python documentation). In practice, though, the line between developers and application users in Python is blurry enough that the only people I've seen actually recommending the use of `DeprecationWarning` in the last few years is core Python developers (see e.g. https://discuss.python.org/t/pendingdeprecationwarning-is-really-useful/1038/18). For all third-party packages `FutureWarning` is generally the safer option since you never really know how your package will be used, unless it's explicitly a command-line tool like click.

In this case I don't think this line adds much either way. I'm going to change it to focus on the important point, which is that `FutureWarning` is shown by default.

File: docs/cudf/source/developer_guide/developer_guide.md
Comment (wence-) [inline docs/cudf/source/developer_guide/developer_guide.md:39]: Does it make sense to flag `pre-commit` up front? This list is informational, but configuring each thing "by hand" is  basically not recommended.

File: docs/cudf/source/developer_guide/developer_guide.md
Comment (wence-) [inline docs/cudf/source/developer_guide/developer_guide.md:8]: Later on you just use `cuDF` rather than "The cuDF library". Maybe start with that here?
```suggestion
cuDF is a GPU-accelerated, [pandas-like](https://pandas.pydata.org/) DataFrame library.
```

(pandas is I think canonically lowercase)

File: docs/cudf/source/developer_guide/developer_guide.md
Comment (wence-) [inline docs/cudf/source/developer_guide/developer_guide.md:27]:
```suggestion
The rest of this document focuses on a high-level overview of best practices in cuDF.
```

**Review by wence- — COMMENTED** [review]
Some style comments/queries

File: docs/cudf/source/developer_guide/developer_guide.md
Comment (mroeschke) [inline docs/cudf/source/developer_guide/developer_guide.md:34]: Coming in cold, probably

1. Layouts that can be enforced by linting rules
2. Layouts that avoid messy dependency chains (`/foo` only imports from `_lib`)

would probably be good start with. In my experience, soft conversions of structure without linting enforcement just tend to stray overtime

File: docs/cudf/source/developer_guide/developer_guide.md
Comment (mroeschke) [inline docs/cudf/source/developer_guide/developer_guide.md:88]: Can add a pre-commit rule for this probably in the future

File: docs/cudf/source/developer_guide/developer_guide.md
Comment (vyasr) [reply docs/cudf/source/developer_guide/developer_guide.md:64]: Hmm that's an interesting question that I haven't considered. Do you think it's worth having a checklist with the items listed here (make sure all calling code is updated, remove tests, etc)?

File: docs/cudf/source/developer_guide/developer_guide.md
Comment (vyasr) [reply docs/cudf/source/developer_guide/developer_guide.md:105]: Ah yes thanks. This requires #11217 to be merged first.

File: docs/cudf/source/developer_guide/library_design.md
Comment (vyasr) [reply docs/cudf/source/developer_guide/library_design.md:215]: So this specific format isn't even just rst, it's [myst](https://myst-parser.readthedocs.io/en/latest/syntax/syntax.html#targets-and-cross-referencing). I'm not sure I follow what you're suggesting, though?

File: docs/cudf/source/developer_guide/developer_guide.md
Comment (vyasr) [reply docs/cudf/source/developer_guide/developer_guide.md:94]: What do you mean by that? Something like "A given version of cuDF will be compatible with the latest version of pandas as of that cuDF release"?

Comment (vyasr) [reply docs/cudf/source/developer_guide/developer_guide.md:94]: CC @shwina for thoughts on what exactly we should promise

Comment (wence-) [reply docs/cudf/source/developer_guide/developer_guide.md:94]: > What do you mean by that? Something like "A given version of cuDF will be compatible with the latest version of pandas as of that cuDF release"?

That's kind of what I had in mind. I guess that the deprecation cycle is probably faster than in pandas (though I don't know)

File: docs/cudf/source/developer_guide/developer_guide.md
Comment (vyasr) [reply docs/cudf/source/developer_guide/developer_guide.md:88]: Ooh good idea.

File: docs/cudf/source/developer_guide/developer_guide.md
Comment (vyasr) [reply docs/cudf/source/developer_guide/developer_guide.md:34]: I don't disagree with you there. This also came up in [the test discussion](https://github.com/rapidsai/cudf/pull/11199#discussion_r918283574), and it's a problem anywhere. Do you have any suggestions on what kind of layouts would be enforceable? I would love that if we could come up with something. CC @shwina

File: docs/cudf/source/developer_guide/developer_guide.md
Comment (vyasr) [reply docs/cudf/source/developer_guide/developer_guide.md:88]: Created #11251

File: docs/cudf/source/developer_guide/developer_guide.md
Comment (vyasr) [reply docs/cudf/source/developer_guide/developer_guide.md]: Our deprecation cycle is typically much faster than pandas is, yes. We don't promise SemVer-style API stability. On the specific topic of pandas compatibility, I really don't think we promise anything other than what I stated in my sentence: each version of cuDF aims for compatibility with the latest version of pandas. That includes edge case behaviors that aren't really API promises in pandas but are things that we test for or users ask us about. I'll let Ashwin chime in on what wording we should use, but I definitely think something like this is worth adding.

File: docs/cudf/source/developer_guide/developer_guide.md
Comment (mroeschke) [reply docs/cudf/source/developer_guide/developer_guide.md:34]: Spitballing simple ones I can think of would be:

1. Mentioned in https://github.com/rapidsai/cudf/pull/11199#discussion_r918283152=, if the doc structure could match test structure (the rst parsing would be a little heavy though)
2. If `cudf/tests/**/` could mirror the structure of `cudf/**` directory without the test directory.

The weakness in both would be where to place integration tests that utilize multiple APIs together.

File: docs/cudf/source/developer_guide/developer_guide.md
Comment (shwina) [reply docs/cudf/source/developer_guide/developer_guide.md]: Perhaps we should only attempt to match behaviour for the *latest" version of Pandas we support? @galipremsagar has much more experience here.

Comment (galipremsagar) [reply docs/cudf/source/developer_guide/developer_guide.md]: > Perhaps we should only attempt to match behavior for the *latest" version of Pandas we support?

+1, that is what we do currently.

File: docs/cudf/source/developer_guide/library_design.md
Comment (wence-) [reply docs/cudf/source/developer_guide/library_design.md:215]: This is a very minor thing, in the grand scheme of things, so I don't think it should hold this PR up, but let me try and explain myself.

> So this specific format isn't even just rst, it's [myst](https://myst-parser.readthedocs.io/en/latest/syntax/syntax.html#targets-and-cross-referencing). 

Myst is just markdown-esque syntax for rst (without the nested parsing issues that docutils have), but this expands into the equivalent of `.. _cythonlayer:` I think.

Here you want to be able to link to this section, but with custom anchors there's no guarantee that someone else won't also pick the same anchor name later.

In this particular case, I think we could turn on [auto-generated anchors](https://myst-parser.readthedocs.io/en/latest/syntax/optional.html#auto-generated-header-anchors) and then the link would be `[](developer_guide/library_design.md#the-cython-layer)`

> I'm not sure I follow what you're suggesting, though? Are you saying that we should have a docs page describing (some parts of) the tool that we use to write the docs?

To me, good online documentation takes advantage of online formats and is link-heavy. If I were writing this in latex, there are good editors that can present overview tables of contents and lists of all section, equation, ... links (with searching). So if I need to reference something, I don't have to remember the name of every section label.

As far as I know, the facilities presented by editors for this in any of the many mark-up languages (including myst-markdown) are quite limited. Consequently, finding links to different sections of the docs requires having them all in my brain (and linking occurs less). [With enough roundtoits I might actually write something for emacs to do this]

The autogenerated header names  have some hierarchy in that they are per-file and the files have a structure (which makes finding things easier), but hand-crafted labels are un-namespaced and so if someone else wants to link they can't guess what to look for is all.

File: docs/cudf/source/developer_guide/developer_guide.md
Comment (wence-) [reply docs/cudf/source/developer_guide/developer_guide.md:65]: I had a look to see how often APIs are deprecated, which is with some regularity. At the moment it doesn't look like there is consistency on whether a tracking issue is created for their subsequent removal. I suspect because that requirement was not previously clearly defined.

While there's a `deprecation` label, it is not consistently applied (there are 24 PRs tagged with deprecation but only 8 issues).

I think an issue template would help with remembering to fill in the issue. I suppose one could even have an automation that enforces that deprecation PRs are linked to by a deprecation issue before merge (but that needs more scripting...).

Maybe something simple like:

```
---
name: Deprecated API tracker
about: Create a tracking issue for removal of a deprecated API
title: "[DEP]"
labels: "deprecation"
assignees: ''

---

** Tracking issue for removal of `FILL_IN_API_CALL`, deprecated in `PR-NUMBER`

- [ ] Calling code removed
- [ ] Tests removed
- [ ] Documentation updated

```

File: docs/cudf/source/developer_guide/developer_guide.md
Comment (wence-) [reply docs/cudf/source/developer_guide/developer_guide.md]: Looks good, thanks.

**Review by mroeschke — APPROVED** [review]
All my comments were addressed.

File: docs/cudf/source/developer_guide/developer_guide.md
Comment (vyasr) [reply docs/cudf/source/developer_guide/developer_guide.md:65]: I am both the person who originally suggested this requirement, and the most frequent violator of it. In hindsight, I am wondering whether this requirement is actually useful. In practice, I have instead been operating using a different strategy. Since our deprecation cycle is a single release, as part of the release cycle I basically do a post-portem search for `FutureWarning` and delete. Do you think that is sufficient, as opposed to the overhead of creating many issues?

CC @shwina @bdice who were involved in the original discussions when I proposed using a label for deprecations.

File: docs/cudf/source/developer_guide/library_design.md
Comment (vyasr) [reply docs/cudf/source/developer_guide/library_design.md:215]: > Myst is just markdown-esque syntax for rst (without the nested parsing issues that docutils have), but this expands into the equivalent of .. _cythonlayer: I think.

That's a reasonable description of Myst, and you're right, that's how a link like this expands. 

Regarding the rest, I 100% agree that both MD and reST documents tend to link far less than they should because of the difficulty of placing, remembering, and using anchors (especially given the variable syntaxes). I would also love better editors. 

> In this particular case, I think we could turn on [auto-generated anchors](https://myst-parser.readthedocs.io/en/latest/syntax/optional.html#auto-generated-header-anchors) and then the link would be [](developer_guide/library_design.md#the-cython-layer)

I hadn't seen this feature. I think that's an awesome idea.

Comment (vyasr) [reply docs/cudf/source/developer_guide/library_design.md:215]: Done.

Comment (wence-) [reply docs/cudf/source/developer_guide/library_design.md]: > I would also love better editors.

Might be worth following progress in https://github.com/swyddfa/esbonio

File: docs/cudf/source/developer_guide/contributing_guide.md
Comment (wence-) [inline docs/cudf/source/developer_guide/contributing_guide.md:60]:
```suggestion
- consist of a single line with no newline characters;
```

File: docs/cudf/source/developer_guide/contributing_guide.md
Comment (isVoid) [inline docs/cudf/source/developer_guide/contributing_guide.md:12]: Wondering if we should mention https://docs.rapids.ai/contributing for Issue, PR basics?

File: docs/cudf/source/developer_guide/developer_guide.md
Comment (bdice) [reply docs/cudf/source/developer_guide/developer_guide.md]: > Since our deprecation cycle is a single release, as part of the release cycle I basically do a post-portem search for `FutureWarning` and delete. Do you think that is sufficient, as opposed to the overhead of creating many issues?

I agree the requirement to create an issue is unnecessary and too much overhead/churn. We sometimes let deprecations slip for 2 releases if we don't have time to rip them out, and that's okay too. We have been in a good habit of keeping things moving from deprecation to removal and I wouldn't over-codify this at present.

Comment (shwina) [reply docs/cudf/source/developer_guide/developer_guide.md]: Agreed with @bdice

**Review by shwina — APPROVED** [review]

File: docs/cudf/source/developer_guide/contributing_guide.md
Comment (vyasr) [reply docs/cudf/source/developer_guide/contributing_guide.md:41]: I like it. I'll adapt that.

File: docs/cudf/source/developer_guide/contributing_guide.md
Comment (vyasr) [reply docs/cudf/source/developer_guide/contributing_guide.md:12]: No, I think if that belongs anywhere it is in the top-level CONTRIBUTING.md.

**Comment by vyasr** [comment]
The status of the guide in this PR is good enough to merge as is with the current set of approvals. Future improvements to developer documentation can always be made without incurring undue additional overhead, so we're better off getting something merged sooner rather than later.

**Comment by vyasr** [comment]
@gpucibot merge

**Pull request merged**
